taigaio/taiga
# feat(attachments, storage): u#3727 t#3961 physical attachment deletion system

<div align="center">
   
![](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExanpoMm9mZG45aXZ6eWVodnZ5cTVkdGFwcWJ1emNxZzQzOW41MTd3eCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/j78GDmi70Lrlkjhc6F/giphy.gif)
</div>   

This PR ensures that the files are deleted when the attachments are deleted. Several things must be taken into consideration:
- Currently, files are saved on disk, but this will change so we will have to use django's storage module.
- In the future we will have to postpone the deletion and make it easily recoverable (to avoid data loss). This is in line with the logical deletion of objects.

The solution that has been reached consists of:
- A new module `taiga.commons.storage` where everything related to file storage will be created.
- A new `StoragedObject` model that will store the information about where the file is stored.
- Now the `Attachment` model will have a `ForeingKey` to `StoragedObject`
- A signal is defined for `Attachment` during the `post_save` that will mark the associated `StoragedObject` as deleted.
- A periodic task is created that will physically delete all files that belong to `StoragedObject` records that have been deleted more than X days ago. (The frequency of the periodic task and time that the files will be kept are configurable parameters).
- There is a CLI command to remove deleted `StoragedObject` items and their files.

## How to tests:

- [x] Tests are green.
- Manual tests: Tests logical deletion:
  - Init:
     - :princess: Open the backend admin panel and access to Storaged Objects list view
     - :bust_in_silhouette: Open the frontend
  - [x]  Tests delete attachment:
     - :bust_in_silhouette: Go to a story and upload some attachments
     - :princess:  refresh and see new storaged objets created
     - :bust_in_silhouette: Delete attachments
     - :princess:  refresh and see storaged objets with their deleted datetime.
   - [x]  Tests delete project:   
      - :bust_in_silhouette: Create a new project.
      - :bust_in_silhouette: Create a new story and upload some attachments (you can create more than one story with attachments)
      - :princess:  refresh and see new storaged objets created 
      - :bust_in_silhouette: Delete the project
      - :princess:  refresh and see all the storaged objets with their deleted datetime.
   -  [x] [Optional] Tests delete user with workspaces and projects with stories with atachments:
      - You can test a more complex scenario: deleting a user involves deleting workspaces that have projects with stories "full" of attachments.      
- Manual tests: Tests physical deletion:
  - [x] Test CLI: `python -m taiga storage clean-storaged-objects --help`  
  - [x] Test periodic task: You can modify `TAIGA_STORAGE__DAYS_TO_STORE_DELETED_STORAGED_OBJECTS` (int or None for all) and `TAIGA_STORAGE__DAYS_TO_STORE_DELETED_STORAGED_OBJECTS` (use [cron format](https://www.freeformatter.com/cron-expression-generator-quartz.html))
